src/hotspot/share/nmt/vmatree.hpp

```diff
@@ -0,0 +1,162 @@
+  public:
+    static int cmp(position a, position b) {
+      if (a < b) return -1;
+      if (a == b) return 0;
+      if (a > b) return 1;
+      ShouldNotReachHere();
+    }
+  };
+
+public:
+  enum class StateType : uint8_t { Reserved, Committed, Released };
+
+  // Each point has some stack and a flag associated with it.
+  struct Metadata {
+    const NativeCallStackStorage::StackIndex stack_idx;
+    const MEMFLAGS flag;
+
+    Metadata() : stack_idx(), flag(mtNone) {}
+
+    Metadata(NativeCallStackStorage::StackIndex stack_idx, MEMFLAGS flag)
+    : stack_idx(stack_idx), flag(flag) {}
+
+    static bool equals(const Metadata& a, const Metadata& b) {
+      return a.flag == b.flag &&
+             NativeCallStackStorage::StackIndex::equals(a.stack_idx, b.stack_idx);
+    }
+  };
+
+  struct IntervalState {
+  private:
+    // Store the type and flag as two bytes
+    uint8_t type_flag[2];
+    NativeCallStackStorage::StackIndex sidx;
+
+  public:
+    IntervalState() : type_flag{0,0}, sidx() {}
+    IntervalState(StateType type, Metadata data) {
+      type_flag[0] = static_cast<uint8_t>(type);
+      type_flag[1] = static_cast<uint8_t>(data.flag);
+      sidx = data.stack_idx;
+    }
+
+    StateType type() const {
+      return static_cast<StateType>(type_flag[0]);
+    }
+
+    MEMFLAGS flag() const {
+      return static_cast<MEMFLAGS>(type_flag[1]);
+    }
+
+    Metadata metadata() const {
+      return Metadata{sidx, flag()};
+    }
+
+    const NativeCallStackStorage::StackIndex stack() const {
+     return sidx;
+    }
+  };
+
+  // An IntervalChange indicates a change in state between two intervals. The incoming state
+  // is denoted by in, and the outgoing state is denoted by out.
+  struct IntervalChange {
+    IntervalState in;
+    IntervalState out;
+
+    bool is_noop() {
+      return (in.type() == StateType::Released && out.type() == StateType::Released) ||
+             (in.type() == out.type() && Metadata::equals(in.metadata(), out.metadata()));
+    }
+  };
+
+  using VMATreap = TreapCHeap<position, IntervalChange, AddressComparator>;
+  using TreapNode = VMATreap::TreapNode;
+
+private:
+  VMATreap _tree;
+
+public:
+  VMATree() : _tree() {}
+
+  struct SingleDiff {
+    int64_t reserve;
+    int64_t commit;
+  };
+  struct SummaryDiff {
+    SingleDiff flag[mt_number_of_types];
+    SummaryDiff() {
+      for (int i = 0; i < mt_number_of_types; i++) {
+        flag[i] = SingleDiff{0, 0};
+      }
+    }
+  };
+
+  SummaryDiff register_mapping(position A, position B, StateType state, Metadata& metadata);
+
+  SummaryDiff reserve_mapping(position from, position sz, Metadata& metadata) {
+    return register_mapping(from, from + sz, StateType::Reserved, metadata);
+  }
+
+  SummaryDiff commit_mapping(position from, position sz, Metadata& metadata) {
+    return register_mapping(from, from + sz, StateType::Committed, metadata);
+  }
+
+  SummaryDiff release_mapping(position from, position sz) {
+    Metadata empty;
```

Comment:
Oh, another thing, maybe rename this to something else. `Metadata` has a clear meaning in hotspot. Maybe something like RegionData?

Reply:
Hi Thomas, what naming scheme should I use for the empty static RegionData? `EmptyRegionData` or `empty_regiondata`? It can't be `constexpr`, but we can make it `const`.

Reply:
If its scoped to RegionData, I would just call it "empty". `RegionData::empty` is clear enough. If its global scope, I'd call it empty_regiondata.